### ISSUE TITLE
Proposal: templates for programs

### ISSUE THREAD

**Issue by Chunosov** [ctuning/ck-autotuning#28]
Currently in order to make a new program we have to make a copy of some existed one to get non-empty meta populated with correct values: compiler dep, run-time entry and so on.
```
ck cp some-repo:program:some-cpp-program some-repo:program:another-cpp-program
```

It would be nice to have some templates, so one could call:
```
ck add some-repo:program:new-cpp-program --template=cpp
```
and get correct meta, initial source file, and may be even readme.

Just after previous command, one could call
```
ck compile new-cpp-program
ck run new-cpp-program
```
and this should be working and producing some dummy output.

**Comment by gfursin** [comment]
Thanks! I have a few ideas how to make it nice via CK and will share thoughts shortly! It should be very useful!

**Comment by gfursin** [comment]
First prototype is done! Now you can make any existing program a template by just adding "template":"yes" to its meta. Then when you add a new program, it will search for such templates and ask you to select one (or create an empty entry). Whenever you select a template, it will create an entry, update meta and will copy all files! It also means that anyone else can add such templates in their own repositories including even private ones!

I created first dedicated template "ck-autotuning:program:template-hello-world-c". Try adding a new program like that:
```
$ ck pull all

$ ck add program:my-new-test
Select 0 (template-hello-world-c)

$ ck find program:my-new-test

$ ck compile program:my-new-test
$ ck run program:my-new-test

$ ck run program:my-new-test --env.CK_VAR1=222
```

You can use a given template immediately as following:
```
$ ck add program:my-new-test2 --template=template-hello-world-c
```

Furthermore, we can now reuse the same functionality in soft, package and dataset - I will try to make first examples later ...

Basically, I created a function "prepare_entry_template" in ck-env:module:misc:
* https://github.com/ctuning/ck-env/blob/master/module/misc/module.py#L340

and I call it from ck-autotuning:module:program in "add" function:
* https://github.com/ctuning/ck-autotuning/blob/master/module/program/module.py#L6909

I think it's really cool - I wanted to add this functionality long time ago but never had time - you motivated me to add it ;) ...

**Comment by gfursin** [comment]
Please, check and tell me if it works ... Comments are welcome!

**Comment by Chunosov** [comment]
Wow! It works!
```
ck add ch-ck-tests:program:my-new-test
0
ck compile program:my-new-test
ck run program:my-new-test
```
```
 (printing output files) 

    * stderr.log

      

    * stdout.log

      Hello world!
      
      CK_VAR1=123
      CK_VAR2=666
```

**Comment by gfursin** [comment]
Cool! Feel free to add new templates:
```
$ ck add ck-autotuning:program:template-XYZ --data_name="some user friendly name"
```
Don't forget to add "template":"yes" and "template" to "tags" list in the meta.json. 
We may need some for python examples and more complex for TensorFlow clients. Later we can add similar templates for TBD benchmark ...

**Comment by gfursin** [comment]
Since it now works, I am closing this ticket!
